base/mpfr.jl

```diff
@@ -889,22 +894,33 @@ isone(x::BigFloat) = x == Clong(1)
 @eval typemax(::Type{BigFloat}) = $(BigFloat(Inf))
 @eval typemin(::Type{BigFloat}) = $(BigFloat(-Inf))
 
-function nextfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
-          z, x, ROUNDING_MODE[])
-    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), z) != 0
-    return z
+function nextfloat!(x::BigFloat, n::Integer)
+    n==0 && return x
+    signbit(n) && return prevfloat!(x, abs(n))
+    for i = 1:n
+        ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x)
+    end
+    return x
 end
 
-function prevfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
-          z, x, ROUNDING_MODE[])
-    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), z) != 0
-    return z
+function prevfloat!(x::BigFloat, n::Integer)
+    n==0 && return x
+    signbit(n) && return nextfloat!(x, abs(n))
+    for i = 1:n
+        ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x)
+    end
+    return x
 end
 
+nextfloat(x::BigFloat, n::Integer) = n==0 ? x : nextfloat!(_duplicate(x), n)
+prevfloat(x::BigFloat, n::Integer) = n==0 ? x : prevfloat!(_duplicate(x), n)
+
+nextfloat!(x::BigFloat) = (ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x); x)
```

Comment:
Why not instead add a default argument, like `nextfloat!(x::BigFloat, n::Integer=1)` ?

Reply:
I think that does not decomplexify; and since `nextfloat!` is being introduced and not exported .. keeping it as simple as possible lets the exported function be more reliably maintained.

Reply:
Well, I proposed this change precisely because that seems simpler to me, since default args are a well understood feature, and this allows to merge two definitions into one, with easier maintainability IMHO.

Reply:
I missed some of your earlier comment.  You are talking about giving `n` the default value of `1`.  I'd have no objection to that in any of the four functions: `nextfloat!`, `prevfloat!`, `nextfloat`, `prevfloat` were it not the case that `float.jl` does not give `nextfloat`, `prevfloat` default values.  I'm sure that's at least in part because `float.jl` was written eariler.  Nonetheless, there is a proclivity to follow the conventions that are in use.

Reply:
> float.jl does not give nextfloat, prevfloat default values.

My guess would be that it's because those functions are documented separately. By the way, the docstring of `nextfloat(x, n)` should be updated here to include `BigFloat` (or simply to remove the type constraint `IEEEFloat`, which is not even exported or documented).

> proclivity

nice new word learnt, thanks!

Reply:
I just checked the code in `float.jl`. It has been implemented like this `nextfloat(x::AbstractFloat) = nextfloat(x,1)`. We can do the same thing here. Also, should I change the rounding mode to global default or use `MPFRRoundNearest` instead?

Reply:
> We can do the same thing here

We can also do it differently, this is what default arguments are for.

Reply:
Why is it a breaking change -- what is the rounding now?

Reply:
Currently, we can do `setrounding(T, Mode)` and the `nexfloat()` uses that Mode. What is implemented in this PR is RoundNearest Mode. 
I would prefer the method mentioned by @rfourquet but would not like to break the code style.

Reply:
Yes, however when there is no `setrounding` set, the default is `RoundNearest`.  So its not clear to me that using `RoundNearest` is breaking anything.  If `setrounding` sets rounding to `RoundDown` or to `RoundUp` what difference does it make -- the nextfloat is the same as the nextfloat rounded down and the same as the nextfloat rounded up.

Reply:
> I would prefer the method mentioned by @rfourquet but would not like to break the code style.

I wonder why you would think that this would break code style. It would rather be your current version which breaks style. If there is no speed advantage (which I doubt in this case) to having two separate bodies with repetition, just merge it into one function with default argument.